### ISSUE TITLE
Question about the test of CIRR

### ISSUE THREAD

**Issue by Code-kunkun** [lucas-ventura/CoVR#15]
<img width="582" alt="image" src="https://github.com/lucas-ventura/CoVR/assets/58265850/3d2f9b0e-2f5e-44a9-ab6e-38fa2c6baba5">

When calculating the similarity between a query and images in CIRR, it seems that the entire image database only includes the reference image mentioned in the triplets, rather than considering all the images in the entire test set. This seems to be somewhat problematic.

**Comment by lucas-ventura** [comment]
Hi @Code-kunkun,

Thank you for bringing this to our attention! You're correct in your observation. I initially assumed that all targets had associated reference images, but it turns out there are 137 targets, approximately 6% of the dataset, without references.

I have addressed this issue in the latest commit (fba3427), so that all images are considered during similarity calculations. You can see the updated results here:
<table>
  <tr>
    <th rowspan="2">Mode</th>
    <th rowspan="2">Evaluation</th>
    <th colspan="4">Recall@K</th>
    <th colspan="3">R_subset@K</th>
  </tr>
  <tr>
    <th>K=1</th>
    <th>K=5</th>
    <th>K=10</th>
    <th>K=50</th>
    <th>K=1</th>
    <th>K=2</th>
    <th>K=3</th>
  </tr>
  <tr>
    <td align="center" rowspan="2">Train (CIRR)</td>
    <td align="center">Old</td>
    <td align="center" colspan="1">49.69</td>
    <td align="center" colspan="1">78.60</td>
    <td align="center" colspan="1">86.77</td>
    <td align="center" colspan="1">94.31</td>
    <td align="center" colspan="1">75.01</td>
    <td align="center" colspan="1">88.12</td>
    <td align="center" colspan="1">93.16</td>
  </tr>
  <tr>
    <td align="center">New</td>
    <td align="center" colspan="1"><span style="font-weight:bold">50.41</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">80.96</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">89.35</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">97.64</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">77.25</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">91.04</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">96.34</span></td>
  </tr>
  <tr>
    <td align="center" rowspan="2">Zero-shot</td>
    <td align="center">Old</td>
    <td align="center" colspan="1">38.48</td>
    <td align="center" colspan="1">66.70</td>
    <td align="center" colspan="1">77.25</td>
    <td align="center" colspan="1">91.47</td>
    <td align="center" colspan="1">69.28</td>
    <td align="center" colspan="1">83.76</td>
    <td align="center" colspan="1">91.11</td>
  </tr>
  <tr>
    <td align="center">New</td>
    <td align="center" colspan="1"><span style="font-weight:bold">39.28</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">68.22</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">78.94</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">94.65</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">71.23</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">86.22</span></td>
    <td align="center" colspan="1"><span style="font-weight:bold">94.02</span></td>
  </tr>
</table>

Thanks again!

Best regards,
Lucas

**Comment by lucas-ventura** [comment]
You're welcome :) 

I've added the functionality to evaluate on the CIRR validation set separately in the latest commit (3b3ab37). You can now run the evaluation using the command: ``python test.py test=cirr-val``

Here are the results I obtained: `{'R1': 29.7536, 'R5': 71.1552, 'R10': 82.4444, 'R50': 95.4556, 'R_mean': 61.1178}`

**Comment by Code-kunkun** [comment]
Thanks for your quick reply!!! However, it seems that the evaluation on the validation set suffers from the same issue as mentioned earlier with the test set, namely, the entire image database does not include all the images from the validation set. I hope the authors can recheck this part of the code, and perhaps provide calculations for Recall@Subset metrics?

**Comment by lucas-ventura** [comment]
Regarding your concern about the validation set, I've checked that all target images in the validation set are indeed included as reference images. Here's a snippet of the code used for this check:
```python
import json
from pathlib import Path

annotation_pth = "annotation/cirr/cap.rc2.val.json"
annotation = json.load(open(annotation_pth, "r"))
ann_ids = {ann["reference"] for ann in annotation}

val_dir = Path("datasets/CIRR/images/dev/")
val_ids = {pth.stem for pth in val_dir.glob("*/*.png")}
```
The result of `len(val_ids - ann_ids)` was 0, confirming that all validation set images are included as references.

Additionally, I’ve implemented the Recall@Subset metrics in the CIRR evaluation code, which I adapted from [CIRPLANT](https://github.com/Cuberick-Orion/CIRPLANT/blob/cd5798dbeb1902aa9b95c323c2f90eb90a757918/model/OSCAR/OSCAR_CIRPLANT.py#L318). 
The changes are in the latest commit (3ed13c5); please let me know if you find any issue.

Best regards,
Lucas

**Comment by Code-kunkun** [comment]
It seems like there's a discrepency between the 2165 elements in ann_ids and 2297 elements in split.rc2.val.json, and the R@1 on CIRR validation set being only 29.75 seems a bit low.

**Comment by lucas-ventura** [comment]
I've updated the code to be similar to the test set (926df9a). I saw that I forgot to normalize the target embeddings, the updated results are the following:
```bash
Recall@1: 41.57
Recall@5: 70.13
Recall@10: 79.91
Recall@50: 94.16
Recall_subset@1: 78.56
Recall_subset@2: 91.73
Recall_subset@3: 95.93
```

**Comment by Code-kunkun** [comment]
Thanks for your reply!!! Here are the results I reproduced on the CIRR validation set, which seems to have some differences from what you provided.

```
Recall@1: 54.58
Recall@5: 82.68
Recall@10: 90.07
Recall@50: 97.58
Recall_subset@1: 78.51
Recall_subset@2: 91.75
Recall_subset@3: 95.93
```

**Comment by lucas-ventura** [comment]
Nice! Which checkpoint are you using? I'm running the following:
```bash
python test.py test=cirr-val model/ckpt=cirr_ft-covr+gt 
```

**Comment by Code-kunkun** [comment]
I used the same command as you 🤔
